pkg/controller/networkpolicy/networkpolicy_controller_test.go

```diff
@@ -15,18 +15,497 @@
+	informerFactory := informers.NewSharedInformerFactory(client, informerDefaultResync)
+	appliedToGroupStore := store.NewAppliedToGroupStore()
+	addressGroupStore := store.NewAddressGroupStore()
+	internalNetworkPolicyStore := store.NewNetworkPolicyStore()
+	npController := NewNetworkPolicyController(client, informerFactory.Core().V1().Pods(), informerFactory.Core().V1().Namespaces(), informerFactory.Networking().V1().NetworkPolicies(), addressGroupStore, appliedToGroupStore, internalNetworkPolicyStore)
+	npController.podListerSynced = alwaysReady
+	npController.namespaceListerSynced = alwaysReady
+	npController.networkPolicyListerSynced = alwaysReady
+	return client, &networkPolicyController{
+		npController,
+		informerFactory.Core().V1().Pods().Informer().GetStore(),
+		informerFactory.Core().V1().Namespaces().Informer().GetStore(),
+		informerFactory.Networking().V1().NetworkPolicies().Informer().GetStore(),
+		appliedToGroupStore,
+		addressGroupStore,
+		internalNetworkPolicyStore,
+	}
+}
+
+func newClientset() *fake.Clientset {
+	client := fake.NewSimpleClientset()
+
+	client.PrependReactor("create", "networkpolicies", k8stesting.ReactionFunc(func(action k8stesting.Action) (bool, runtime.Object, error) {
+		np := action.(k8stesting.CreateAction).GetObject().(*networkingv1.NetworkPolicy)
+
+		if np.ObjectMeta.GenerateName != "" {
+			np.ObjectMeta.Name = fmt.Sprintf("%s-%s", np.ObjectMeta.GenerateName, rand.String(8))
+			np.ObjectMeta.GenerateName = ""
+		}
+
+		return false, np, nil
+	}))
+
+	return client
+}
+
+func TestAddNetworkPolicy(t *testing.T) {
+	ns := metav1.NamespaceDefault
+	npName := "testing-1"
+	pSelector := metav1.LabelSelector{}
+	apgID := getNormalizedUID(generateNormalizedName(ns, &pSelector, nil))
+	npObj := &networkingv1.NetworkPolicy{
+		ObjectMeta: metav1.ObjectMeta{Name: npName, Namespace: ns},
+		Spec: networkingv1.NetworkPolicySpec{
+			PodSelector: pSelector,
+		},
+	}
+	nSelector1 := getK8sLabelSelector(1)
+	nSelectorPeer1 := []networkingv1.NetworkPolicyPeer{
+		{
+			NamespaceSelector: &nSelector1,
+		},
+	}
+	nSelector2 := getK8sLabelSelector(2)
+	nSelectorPeer2 := []networkingv1.NetworkPolicyPeer{
+		{
+			NamespaceSelector: &nSelector2,
+		},
+	}
+	pSelector1 := getK8sLabelSelector(1)
+	pSelectorPeer1 := []networkingv1.NetworkPolicyPeer{
+		{
+			PodSelector: &pSelector1,
+		},
+	}
+	pSelector2 := getK8sLabelSelector(2)
+	pSelectorPeer2 := []networkingv1.NetworkPolicyPeer{
+		{
+			PodSelector: &pSelector2,
+		},
+	}
+	tables := []struct {
+		ir               []networkingv1.NetworkPolicyIngressRule
+		er               []networkingv1.NetworkPolicyEgressRule
+		expRules         int
+		expInAddrGroups  int
+		expOutAddrGroups int
+	}{
+		// Test case with nsSelector Egress rules only.
+		{
+			[]networkingv1.NetworkPolicyIngressRule{},
+			[]networkingv1.NetworkPolicyEgressRule{
+				{
+					To: nSelectorPeer1,
+				},
+				{
+					To: nSelectorPeer2,
+				},
+			},
+			2,
+			0,
+			1,
```

Comment:
There are different selectors in the two egress rules, I assume the test should verify that 2 unique groups will be created? and there should be another test verify that 2 rules that have same selector result in only one unique group?
Is the verification just checking each rule has one group? it seems not covering much logic.

Reply:
FYI, since the default-allow and default-deny logic is moved to controller side, I have updated processNetworkPolicy and added some tests to cover the default-allow, default-deny, rules with same selectors and rules with different selectors there.